### PR TITLE
fix: upgrade to new gateway and federation

### PR DESCRIPTION
**Pull request by marco-ippolito**
Resolves: https://github.com/mercurius-js/cache/issues/126
I've updated to mercurius 12.
I've bumped the version to add support for the new packages mercurius-federation and mercurius-gateway.
I've put this line under coverage ignore:
```
  // istanbul ignore next
  if (app.graphqlGateway) {
    // Add hook to regenerate the resolvers when the schema is refreshed
    app.graphql.addHook('onGatewayReplaceSchema', async (instance, schema) => {
      buildCache()
      setupSchema(schema, policy, all, cache, skip, onDedupe, onHit, onMiss, onSkip, onError, report)
    })
  }
```
since I couldnt come up with a meaningful test for ` onGatewayReplaceSchema ` (the one on gateway is very flaky), open to suggestion

### CODE REVIEW

**Comment by simone-sanfratello** [comment]
Happy to work on a solution for test, not going to accept code without

**Comment by mcollina** [comment]
The ignore should not be needed, because that if should be covered by the gateway test. If it's not then there is a bug.

**Comment by marco-ippolito** [comment]
it is not covered and I tried some basic test replacing schema on the gateway but its still nothing

**Comment by simone-sanfratello** [comment]
From our check, looks like `onGatewayReplaceSchema` hook is not been triggered

**Comment by mcollina** [comment]
It might be a problem in mercurius gateway then

**Comment by marco-ippolito** [comment]
Ill try to remove it also I dont see why someone should cache the `gatewayReplaceSchema` as its not something called often but I might been missing something

**Comment by simone-sanfratello** [comment]
I think we should update the cache on schema change, for example using `all` option I expect all the queries to be updated too

**Comment by marco-ippolito** [comment]
I've added a test that shows that `gatewatOnReplaceSchema` it's not triggered by replace schema on a service. @mcollina might be a bug from the gateway

File: index.js [hunk 1]
```diff
@@ -35,12 +35,14 @@ module.exports = fp(async function (app, opts) {
...
-
-  // Add hook to regenerate the resolvers when the schema is refreshed
-  app.graphql.addHook('onGatewayReplaceSchema', async (instance, schema) => {
-    buildCache()
-    setupSchema(schema, policy, all, cache, skip, onDedupe, onHit, onMiss, onSkip, onError, report)
-  })
+  // istanbul ignore next
+  if (app.graphqlGateway) {
+    // Add hook to regenerate the resolvers when the schema is refreshed
+    app.graphql.addHook('onGatewayReplaceSchema', async (instance, schema) => {
```
Comment (mcollina) [inline index.js:41]:
```suggestion
    app.graphqlGateway.addHook('onGatewayReplaceSchema', async (instance, schema) => {
```

Comment (mcollina) [reply index.js:41]: this should make it triggered.

**Comment by mcollina** [comment]
Is this still wip? @marco-ippolito could you update the PR title if it's ready to go?

I think this is good to land.

@simone-sanfratello PTAL.

**Comment by marco-ippolito** [comment]
Its ready

**Review by simone-sanfratello — APPROVED** [review]
LGTM! :rocket:

**Pull request merged**